### PR TITLE
[Apple TV]  Tools and dependencies support

### PR DESCRIPTION
**Pull request by sy6sy2**
## Description
This PR is the first step in order to add full Apple TV support for Kodi.

This PR only contains minimal modification and addition in order to be able to:

* Build tools and dependencies for Apple TV platform
* Generate Xcode project of Kodi for Apple TV platform

Apple TV update for Kodi core is still in progress.

Forum thread: https://forum.kodi.tv/showthread.php?tid=312805

Contributors: @kambala-decapitator, @fuzzard, @phunkyfish

## Motivation and Context
Add Apple TV support for kodi 18.

The main idea is to port the work done for Kodi 17 by @Memphiz for tvOS (https://github.com/Memphiz/xbmc/tree/yab) to Kodi 18/master

## How Has This Been Tested?
Tested by contributors by following tvOS documentation found in this PR.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [X] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: cmake/scripts/tvos/ArchSetup.cmake [hunk 1]
```diff
@@ -0,0 +1,59 @@
+if(NOT CMAKE_TOOLCHAIN_FILE)
+  message(FATAL_ERROR "CMAKE_TOOLCHAIN_FILE required for tvos. See ${CMAKE_SOURCE_DIR}/cmake/README.md")
+endif()
+
+set(CORE_MAIN_SOURCE ${CMAKE_SOURCE_DIR}/xbmc/platform/darwin/tvos/MainApplication.mm)
+
+set(ARCH_DEFINES -D_LINUX -DTARGET_POSIX -DTARGET_DARWIN -DTARGET_DARWIN_TVOS)
```
Comment (Memphiz) [inline cmake/scripts/tvos/ArchSetup.cmake]: Those are in defined.txt too - not sure if this is an intended duplication

**Review by Memphiz — COMMENTED** [review]
Looks good for me at a glance.
To get this in we need to add tvOS to Jenkins before and have leia branches off so 19 master is free.

**Comment by phunkyfish** [comment]
To clarify, do you mean to get this in before the Leia branch?

**Comment by Memphiz** [comment]
@punkyfish - no I mean the opposite. This is v19 material ...

**Comment by phunkyfish** [comment]
So Leia branches and after that we get this in and add to Jenkins. Make sense. Ty

**Comment by phunkyfish** [comment]
Looks like the Leia branch has been created.

File: cmake/treedata/tvos/subdirs.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+xbmc/platform/linux/network           platform/linux/network
+xbmc/input/touch                      input/touch
+xbmc/input/touch/generic              input/touch/generic
+xbmc/platform/posix                   posix
+xbmc/platform/posix/filesystem        platform/posix/filesystem
+xbmc/platform/posix/utils             platform/posix/utils
+xbmc/platform/darwin                  platform/darwin
+xbmc/platform/darwin/tvos             platform/tvos
+xbmc/platform/darwin/ios-common       platform/ios-common
+xbmc/platform/darwin/osx/network      platform/osx/network
```
Comment (Rechi) [inline cmake/treedata/tvos/subdirs.txt]: tvOS isn't allowed to use osx code. If it is need it has to be moved into a directory named somthing like darwin.

File: cmake/treedata/tvos/subdirs.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+xbmc/platform/darwin/osx/peripherals  platform/osx/peripherals
```
Comment (Rechi) [inline cmake/treedata/tvos/subdirs.txt]: tvOS isn't allowed to use osx code. If it is need it has to be moved into a directory named somthing like darwin.

File: cmake/treedata/tvos/subdirs.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+xbmc/platform/darwin/osx/powermanagement platform/darwin/osx/powermanagement
```
Comment (Rechi) [inline cmake/treedata/tvos/subdirs.txt]: tvOS isn't allowed to use osx code. If it is need it has to be moved into a directory named somthing like darwin.

File: cmake/treedata/tvos/subdirs.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+xbmc/platform/darwin/osx/storage      platform/osx/storage
```
Comment (Rechi) [inline cmake/treedata/tvos/subdirs.txt]: tvOS isn't allowed to use osx code. If it is need it has to be moved into a directory named somthing like darwin.

File: cmake/treedata/tvos/subdirs.txt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+xbmc/windowing/osx                    windowing/osx
```
Comment (Rechi) [inline cmake/treedata/tvos/subdirs.txt]: tvOS has it's own windowing so move it into tvos folder.

File: tools/darwin/packaging/ios/mkdeb-ios.sh.in [hunk 7]
```diff
@@ -83,15 +83,16 @@ mkdir -p $DIRNAME/$PACKAGE/DEBIAN
 if $ARM64; then
   echo "Package: $PACKAGE_ARM64"                  >  $DIRNAME/$PACKAGE/DEBIAN/control
   echo "Name: @APP_NAME@-iOS (64-bit)"            >> $DIRNAME/$PACKAGE/DEBIAN/control
+  echo "Depends: firmware (>= 7.0)"               >> $DIRNAME/$PACKAGE/DEBIAN/control
```
Comment (Rechi) [inline tools/darwin/packaging/ios/mkdeb-ios.sh.in]: Has to be removed.

File: tools/darwin/packaging/ios/mkdeb-ios.sh.in [hunk 8]
```diff
@@ -83,15 +83,16 @@ mkdir -p $DIRNAME/$PACKAGE/DEBIAN
...
   echo "Pre-Depends: cy+cpu.arm64"                >> $DIRNAME/$PACKAGE/DEBIAN/control
   echo "Conflicts: $PACKAGE"                      >> $DIRNAME/$PACKAGE/DEBIAN/control
   echo "Replaces: $PACKAGE"                       >> $DIRNAME/$PACKAGE/DEBIAN/control
 else
   echo "Package: $PACKAGE"                        >  $DIRNAME/$PACKAGE/DEBIAN/control
   echo "Name: @APP_NAME@-iOS"                     >> $DIRNAME/$PACKAGE/DEBIAN/control
+  echo "Depends: firmware (>= 5.1)"               >> $DIRNAME/$PACKAGE/DEBIAN/control
```
Comment (Rechi) [inline tools/darwin/packaging/ios/mkdeb-ios.sh.in]: Has to be removed.

File: tools/darwin/packaging/ios/mkdeb-ios.sh.in [hunk 9]
```diff
@@ -83,15 +83,16 @@ mkdir -p $DIRNAME/$PACKAGE/DEBIAN
...
 fi
 echo "Priority: Extra"                            >> $DIRNAME/$PACKAGE/DEBIAN/control
-echo "Depends: firmware (>= 9.0)"                 >> $DIRNAME/$PACKAGE/DEBIAN/control
```
Comment (Rechi) [inline tools/darwin/packaging/ios/mkdeb-ios.sh.in]: Has to be removed.

File: tools/depends/configure.ac [hunk 10]
```diff
@@ -320,45 +320,65 @@ case $host in
         MC_CHECK_NOT_CPU([$use_cpu], "*86")
 
         # setup which sdk to use
+
+        # If tvOS
         if test "$use_platform" = "tvos"; then
+          platform_os="tvos"
```
Comment (Rechi) [inline tools/depends/configure.ac]: Why make tvOS a complete separate platform instead of a special variant of the iOS platform, like it was done before?

File: tools/depends/configure.ac [hunk 11]
```diff
@@ -419,6 +441,8 @@ case $use_platform in
      platform_cxxflags="-fPIC -mcpu=cortex-a53 -mfloat-abi=hard -mfpu=neon-fp-armv8 -mvectorize-with-neon-quad"
      platform_ldflags="-lpthread"
      ;;
+  tvos)
+    ;;
```
Comment (Rechi) [inline tools/depends/configure.ac]: Needs a `platform_os` check like aml, gbm and wayland.

File: tools/depends/native/cmake-native/Makefile [hunk 12]
```diff
@@ -47,4 +47,3 @@ clean:
 
 distclean::
 	rm -rf $(PLATFORM) .installed-$(PLATFORM)
-
```
Comment (Rechi) [inline tools/depends/native/cmake-native/Makefile]: unrelated

File: tools/depends/target/Toolchain.cmake.in [hunk 13]
```diff
@@ -31,24 +31,48 @@ elseif(OS STREQUAL osx)
   set(CMAKE_SYSTEM_NAME Darwin)
   set(CORE_SYSTEM_NAME osx)
 elseif(OS STREQUAL ios)
-  set(CMAKE_SYSTEM_NAME Darwin)
-  set(CORE_SYSTEM_NAME ios)
+    set(CMAKE_SYSTEM_NAME Darwin)
+    set(CORE_SYSTEM_NAME ios)
```
Comment (Rechi) [inline tools/depends/target/Toolchain.cmake.in]: unrelated indentation change

File: tools/depends/target/Toolchain.cmake.in [hunk 14]
```diff
@@ -31,24 +31,48 @@ elseif(OS STREQUAL osx)
...
+    # Necessary to build the main Application (but not other dependencies)
+    # with Xcode (and a bundle with Makefiles) (https://cmake.org/Bug/view.php?id=15329)
+    if(NOT PROJECT_SOURCE_DIR MATCHES "tools/depends")
+      message(STATUS "Toolchain enabled IOS bundle for project ${PROJECT_NAME}")
+      set(CMAKE_MACOSX_BUNDLE YES)
+      set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGNING_REQUIRED "NO")
+
+      # set this to YES once we have a deployment target of at least iOS 6.0
+      set(CMAKE_XCODE_ATTRIBUTE_ENABLE_BITCODE "NO")
+    endif()
```
Comment (Rechi) [inline tools/depends/target/Toolchain.cmake.in]: unrelated indentation change

**Comment by SylvainCecchetto** [comment]
Last force push:
* Add tvOS doc section in docs/README
* Add clean/distclean in binary-addons MakeFile

File: tools/depends/target/Toolchain.cmake.in (hunk 14 above)
Comment (SylvainCecchetto) [reply tools/depends/target/Toolchain.cmake.in]: Done

File: tools/depends/target/Toolchain.cmake.in (hunk 13 above)
Comment (SylvainCecchetto) [reply tools/depends/target/Toolchain.cmake.in]: Done

File: tools/depends/configure.ac (hunk 10 above)
Comment (SylvainCecchetto) [reply tools/depends/configure.ac]: Memphiz thought it was a good idea to separate thing concerning iOS and tvOS (ref: https://forum.kodi.tv/showthread.php?tid=312805&pid=2842516#pid2842516)

Comment (phunkyfish) [reply tools/depends/configure.ac]: @Memphiz, do you have a preference on setting the OS to tvos or leaving it as ios and using target platform to differentiate instead (similar to your yab branch).

Comment (Memphiz) [reply tools/depends/configure.ac]: Well I have no idea how similar all this is on the cmake side. I would have split cmake side because iOS and tvOS have different platform implementations for entry point and controller and some other classes. I’d say go for the route with lesser if deferi needed. Maybe @Rechi has some more input why separation is not a good idea here.
iOS is a touch device os and tvOS is a set top box os while there are some similarities there also a lot of differences.

Comment (Rechi) [reply tools/depends/configure.ac]: The reason I'm asking is the following comment from `TargetConditionals.h` in the platforms SDKs. Also most of the build system code looks the same. I was thinking to handle it like the different windowing for linux. 
```
TARGET_OS_* 
These conditionals specify in which Operating System the generated code will
run.  Indention is used to show which conditionals are evolutionary subclasses.  

The MAC/WIN32/UNIX conditionals are mutually exclusive.
The IOS/TV/WATCH conditionals are mutually exclusive.


    TARGET_OS_WIN32           - Generated code will run under 32-bit Windows
    TARGET_OS_UNIX            - Generated code will run under some Unix (not OSX) 
    TARGET_OS_MAC             - Generated code will run under Mac OS X variant
       TARGET_OS_OSX          - Generated code will run under OS X devices
       TARGET_OS_IPHONE          - Generated code for firmware, devices, or simulator
          TARGET_OS_IOS             - Generated code will run under iOS 
          TARGET_OS_TV              - Generated code will run under Apple TV OS
          TARGET_OS_WATCH           - Generated code will run under Apple Watch OS
          TARGET_OS_BRIDGE          - Generated code will run under Bridge devices
       TARGET_OS_SIMULATOR      - Generated code will run under a simulator
       TARGET_OS_EMBEDDED       - Generated code for firmware
   
    TARGET_IPHONE_SIMULATOR   - DEPRECATED: Same as TARGET_OS_SIMULATOR
    TARGET_OS_NANO            - DEPRECATED: Same as TARGET_OS_WATCH

  +------------------------------------------------+
  |                TARGET_OS_MAC                   |
  | +---+  +-------------------------------------+ |
  | |   |  |          TARGET_OS_IPHONE           | |
  | |OSX|  | +-----+ +----+ +-------+ +--------+ | |
  | |   |  | | IOS | | TV | | WATCH | | BRIDGE | | |
  | |   |  | +-----+ +----+ +-------+ +--------+ | |
  | +---+  +-------------------------------------+ |
  +------------------------------------------------+
 ```

Comment (SylvainCecchetto) [reply tools/depends/configure.ac]: Thank you this information. I understand better this questioning.

We revert changes to have the most common things between iOS and tvOS.

**Review by Rechi — COMMENTED** [review]
Commit order is wrong, looks like it's nearly reveres the expected order.
- configure.ac change to allow `with-platform=tvos`
- depends changes (tools/depends/target) or if you like combined in the first commit
- cmake, core Kodi changes, ... 

The reason it should be this way is, without the config change the other changes are useless and so on.

**Comment by SylvainCecchetto** [comment]
Thank you for your review.

I deleted a lot of things in order to have a smaller PR (normally all `cmake/*` changes are only mandatory to build Kodi core so we will focus on theses changes in a later PR with core changes).
So this PR only add the ability to build dependencies for tvOS and to generate the Xcode project with the toolchain.

We still have random errors when using make in multi threaded mode but note sure if it's related to tvOS changes (see https://github.com/SylvainCecchetto/xbmc/issues/23). 

I am currently trying to build dependencies from scratch with this branch state  to see if nothing is broken with all removed changes.

Do you know why jenkins sometime fails to start building?

File: tools/depends/configure.ac [hunk 15]
```diff
@@ -346,19 +361,22 @@ case $host in
...
+            11.*);;
+            12.*);;
             *)
               AC_MSG_ERROR(error in configure of --with-sdk=$use_sdk)
-            ;;
-        esac
+              ;;
+          esac
+
+        fi
```
Comment (Rechi) [inline tools/depends/configure.ac]: I would avoid duplicating the code which is valid for iOS and tvOS, see https://github.com/xbmc/xbmc/compare/master...Rechi:depends/appletvos.

File: tools/depends/configure.ac [hunk 16]
```diff
@@ -679,13 +703,12 @@ AC_SUBST(ffmpeg_options)
...
-  if test "$use_platform" = "tvos"; then
-    simulator_sdk_path=[`$use_xcodebuild -version -sdk appletvsimulator$use_sdk  | grep ^Path | awk '{ print $2}'`]
-  else
+  if test "$target_platform" == "iphoneos"; then
     simulator_sdk_path=[`$use_xcodebuild -version -sdk iphonesimulator$use_sdk  | grep ^Path | awk '{ print $2}'`]
+    cp -vf $simulator_sdk_path/usr/include/crt_externs.h $prefix/$deps_dir/include
   fi
-  cp -vf $simulator_sdk_path/usr/include/crt_externs.h $prefix/$deps_dir/include
 fi
+
```
Comment (Rechi) [inline tools/depends/configure.ac]: Why is this change needed?

File: tools/depends/target/Toolchain.cmake.in [hunk 17]
```diff
@@ -32,23 +32,37 @@ elseif(OS STREQUAL osx)
   set(CORE_SYSTEM_NAME osx)
 elseif(OS STREQUAL ios)
   set(CMAKE_SYSTEM_NAME Darwin)
-  set(CORE_SYSTEM_NAME ios)
+  if(PLATFORM STREQUAL appletvos)
+    set(CORE_SYSTEM_NAME tvos)
```
Comment (Rechi) [inline tools/depends/target/Toolchain.cmake.in]: Remove this from the PR as the code which may need it is not added in this PR.

File: tools/depends/target/Toolchain.cmake.in [hunk 18]
```diff
@@ -32,23 +32,37 @@ elseif(OS STREQUAL osx)
...
+    elseif(CORE_SYSTEM_NAME STREQUAL tvos)
+     message(STATUS "Toolchain enabled TVOS bundle for project ${PROJECT_NAME}")
+     # Need to set this attribute to "" in order to
+     # complety disable code signing
+     # see: https://gitlab.kitware.com/cmake/cmake/issues/19112
+     set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY "")
+
+     # set this to YES once we have a deployment target of at least iOS 6.0
+     set(CMAKE_XCODE_ATTRIBUTE_ENABLE_BITCODE "YES")
+    endif()
```
Comment (Rechi) [inline tools/depends/target/Toolchain.cmake.in]: Further commands can be moved out of the if.

File: tools/depends/target/openssl/Makefile [hunk 19]
```diff
@@ -43,7 +49,11 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 		sed -E -ie "s|^CFLAGS=-|CFLAGS=$(CFLAGS) -|" "$(PLATFORM)/Makefile"; \
 		sed -ie "s|-isysroot \$$(CROSS_TOP)/SDKs/\$$(CROSS_SDK) ||" "$(PLATFORM)/Makefile"; \
 		sed -ie "s|static volatile sig_atomic_t intr_signal;|static volatile intr_signal;|" "$(PLATFORM)/crypto/ui/ui_openssl.c"; \
+		if test "$(TARGET_PLATFORM)" = "appletvos"; then \
+			sed -i -- 's/define HAVE_FORK 1/define HAVE_FORK 0/' "$(PLATFORM)/apps/speed.c"; \
+		fi; \
 	fi
+
 	sed -ie "s|apps test||" "$(PLATFORM)/Makefile"; \
```
Comment (Rechi) [inline tools/depends/target/openssl/Makefile]: Fixing this command will fix the fork error, see https://github.com/xbmc/xbmc/compare/master...Rechi:depends/appletvos. Updating this step was missed in the OpenSSL 1.1.0 bump.

File: tools/depends/target/Toolchain.cmake.in (hunk 17 above)
Comment (SylvainCecchetto) [reply tools/depends/target/Toolchain.cmake.in]: Toolchain file removes.
You are right; this file is not needed to build deps.
We keep theses changes for a future PR concerning Kodi core.

File: tools/depends/configure.ac (hunk 15 above)
Comment (SylvainCecchetto) [reply tools/depends/configure.ac]: I copy paste your changes, thank you.

File: tools/depends/configure.ac (hunk 16 above)
Comment (SylvainCecchetto) [reply tools/depends/configure.ac]: Reverted like master state.

File: tools/depends/target/openssl/Makefile (hunk 19 above)
Comment (SylvainCecchetto) [reply tools/depends/target/openssl/Makefile]: Thank you. Modification done.

**Review by Rechi — APPROVED** [review]

**Comment by Memphiz** [comment]
built successful on tvOS (depends only) https://jenkins.kodi.tv/job/TVOS/36/

**Comment by Rechi** [comment]
jenkins build this please

**Comment by Memphiz** [comment]
@Rechi please merge once jenkins is happy ...
